chia/full_node/mempool_check_conditions.py

```diff
@@ -132,40 +134,222 @@ def mempool_assert_my_amount(condition: ConditionWithArgs, unspent: CoinRecord)
-            cost, result = GENERATOR_MOD.run_safe_with_cost(max_cost, block_program, block_program_args)
+            clvm_cost, result = GENERATOR_MOD.run_safe_with_cost(max_cost, block_program, block_program_args)
         else:
-            cost, result = GENERATOR_MOD.run_with_cost(max_cost, block_program, block_program_args)
+            clvm_cost, result = GENERATOR_MOD.run_with_cost(max_cost, block_program, block_program_args)
+
+        max_cost -= clvm_cost
+        if max_cost < 0:
+            return NPCResult(uint16(Err.INVALID_BLOCK_COST.value), [], uint64(0))
         npc_list: List[NPC] = []
-        opcodes: Set[bytes] = set(item.value for item in ConditionOpcode)
 
         for res in result.first().as_iter():
             conditions_list: List[ConditionWithArgs] = []
 
+            if len(res.first().atom) != 32:
+                raise ValidationError(Err.INVALID_CONDITION)
             spent_coin_parent_id: bytes32 = res.first().as_atom()
-            spent_coin_puzzle_hash: bytes32 = res.rest().first().as_atom()
-            spent_coin_amount: uint64 = uint64(res.rest().rest().first().as_int())
+            res = res.rest()
+            if len(res.first().atom) != 32:
+                raise ValidationError(Err.INVALID_CONDITION)
+            spent_coin_puzzle_hash: bytes32 = res.first().as_atom()
+            res = res.rest()
+            spent_coin_amount: uint64 = uint64(res.first().as_int())
+            res = res.rest()
             spent_coin: Coin = Coin(spent_coin_parent_id, spent_coin_puzzle_hash, spent_coin_amount)
 
-            for cond in res.rest().rest().rest().first().as_iter():
-                if cond.first().as_atom() in opcodes:
-                    opcode: ConditionOpcode = ConditionOpcode(cond.first().as_atom())
-                elif not safe_mode:
-                    opcode = ConditionOpcode.UNKNOWN
-                else:
-                    return NPCResult(uint16(Err.GENERATOR_RUNTIME_ERROR.value), [], uint64(0))
-                cvl = ConditionWithArgs(opcode, cond.rest().as_atom_list())
-                conditions_list.append(cvl)
+            for cond in res.first().as_iter():
+                cost, cvl = parse_condition(cond, safe_mode)
+                max_cost -= cost
+                if max_cost < 0:
+                    return NPCResult(uint16(Err.INVALID_BLOCK_COST.value), [], uint64(0))
+                if cvl is not None:
+                    conditions_list.append(cvl)
+
             conditions_dict = conditions_by_opcode(conditions_list)
             if conditions_dict is None:
                 conditions_dict = {}
             npc_list.append(
                 NPC(spent_coin.name(), spent_coin.puzzle_hash, [(a, b) for a, b in conditions_dict.items()])
             )
-        return NPCResult(None, npc_list, uint64(cost))
+        return NPCResult(None, npc_list, uint64(clvm_cost))
+    except ValidationError as e:
+        return NPCResult(uint16(e.code.value), [], uint64(0))
```

Comment:
this is how the `parse_*()` functions produce the correct error code

Reply:
Using `ValidationError` is a better idea than returning the error explicitly, in my opinion... you might actually want to refactor to a new function that doesn't return a triple (which is really a `Union` in disguise) but fails with a `ValidationError` raise, and wrap this old legacy API in it. Sort of inching things slowly into a more standard direction.